src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Manages properties of analyzers (such as registered actions, supported diagnostics) for analyzer host's lifetime.
+    /// 
+    /// It ensures the following for the lifetime of analyzer host:
+    /// 1) <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> is invoked only once per-analyzer.
+    /// 2) <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is invoked only once per-analyzer.
+    /// 3) <see cref="CompilationStartAnalyzerAction"/> registered during Initialize are invoked only once per-analyzer per-compilation.
+    /// </summary>
+    internal class AnalyzerManager
+    {
+        public static readonly AnalyzerManager Default = new AnalyzerManager();
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
+
+        /// <summary>
+        /// Cache descriptors for each diagnostic analyzer. We do this since <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is
+        /// a property rather than metadata. We expect it to be cheap and immutable, but we can't force them to be so, we cache them
+        /// and ask only once.
+        /// </summary>
+        private readonly ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>> _descriptorCache =
```

Comment:
because we can't control lifetime of DiagnosticAnalyzer. if DiagnosticAnalyzer is ever held strongly, there should be something manage lifetime of DiagnosticAnalyzer. like subscribing to AnalyzerReference changed event or something. otherwise, we will never let DiagnosticAnalyzer once created go away. basically memory leaks.

I didn't want to manage it explicitly, that is why I used ConditionalWeakTable.

one question though, what is lifetime expectation for AnalyzerDriver you are creating? I thought its lifetime is same AnalyzerDriver. but sounds like it is not, then.. not sure how you can strongly hold on to DiagnosticAnalyzer.